packages/framework/src/Model/Product/Recalculation/ProductRecalculationFacade.php

```diff
@@ -34,14 +38,75 @@ public function __construct(
     {
-        $this->productVisibilityFacade->calculateProductVisibilityForIds($productIds);
+        $idsToRecalculate = $this->getIdsToRecalculate($productIds);
+
+        $this->productVisibilityFacade->calculateProductVisibilityForIds($idsToRecalculate);
 
         foreach ($this->domain->getAllIds() as $domainId) {
             $this->indexFacade->exportIds(
                 $this->indexRegistry->getIndexByIndexName(ProductIndex::getName()),
                 $this->indexDefinitionLoader->getIndexDefinition(ProductIndex::getName(), $domainId),
-                $productIds,
+                $idsToRecalculate,
             );
         }
     }
+
+    /**
+     * @param array $productIds
+     * @return array
+     */
+    protected function getIdsToRecalculate(array $productIds): array
+    {
+        $mainVariantsIds = $this->getMainVariantIdsOfVariants($productIds);
+
+        $productIdsWithAddedMainVariantIds = [...$productIds, ...$mainVariantsIds];
+
+        $regularProductOrMainVariantIds = $this->dropVariantIdsFromProductIds($productIdsWithAddedMainVariantIds);
+        $variantIds = $this->getVariantIds($productIdsWithAddedMainVariantIds);
+
+        return [...$variantIds, ...$regularProductOrMainVariantIds];
+    }
+
+    /**
+     * @param int[] $productIds
+     * @return int[]
+     */
+    protected function getMainVariantIdsOfVariants(array $productIds): array
+    {
+        $result = $this->em->createQuery('SELECT IDENTITY(p.mainVariant) as id FROM ' . Product::class . ' p WHERE p.id IN (:productIds) AND p.variantType = :variantTypeVariant')
+            ->setParameter('productIds', $productIds)
+            ->setParameter('variantTypeVariant', Product::VARIANT_TYPE_VARIANT)
+            ->getResult();
+
+        return array_column($result, 'id');
+    }
+
+    /**
+     * @param int[] $productIds
+     * @return int[]
+     */
+    protected function dropVariantIdsFromProductIds(array $productIds): array
+    {
+        $result = $this->em->createQuery('SELECT p.id FROM ' . Product::class . ' p WHERE p.id IN (:productIds) AND (p.variantType = :variantTypeMain OR p.variantType = :variantTypeNone)')
+            ->setParameter('productIds', $productIds)
+            ->setParameter('variantTypeMain', Product::VARIANT_TYPE_MAIN)
+            ->setParameter('variantTypeNone', Product::VARIANT_TYPE_NONE)
+            ->getResult();
+
+        return array_column($result, 'id');
+    }
+
+    /**
+     * @param int[] $mainVariantsIds
+     * @return int[]
+     */
+    protected function getVariantIds(array $mainVariantsIds): array
+    {
+        $result = $this->em->createQuery('SELECT p.id FROM ' . Product::class . ' p WHERE p.mainVariant IN (:productIds) AND p.variantType = :variantTypeVariant')
+            ->setParameter('productIds', $mainVariantsIds)
+            ->setParameter('variantTypeVariant', Product::VARIANT_TYPE_VARIANT)
+            ->getResult();
+
+        return array_column($result, 'id');
```

Comment:
these methods fit a little bit more into `ProductRepository` IMHO, but I do not insist on that

Reply:
maybe a new repository for the new feature? 😇

Reply:
I have added a new repository :) it's now more readable, thanks